### ISSUE TITLE
Default python variant +libxml2 leads to dependency cycle when installing itstool or libxslt or dependents

### ISSUE THREAD

**Issue by rkevk** [LLNL/spack#34963]
### Steps to reproduce the issue

When installing `py-matplotlib` with the `+latex` or `+animation` variant, both `python~libxml2` and `python+libxml2` are simultaneously listed as dependencies. This propagates down the python dependency tree and can cause packages like `py-numpy`, `py-cython` etc. to be needlessly reinstalled. 
Furthermore, at runtime (following `spack load py-matplotlib`) both competing python builds are loaded simultaneously!

I did not check in detail how each combination of the different `matplotlib` variants might cause to differing behavior.

Spec files:
[spack-spec+animation.txt](https://github.com/spack/spack/files/10428565/spack-spec%2Banimation.txt)
[spack-spec+latex.txt](https://github.com/spack/spack/files/10428566/spack-spec%2Blatex.txt)

### Error message

No error messages, but needlessly installing two competing sets of software is undesired behavior.

However, note that: 

- Adding a `^python libxml2==True` to the `+animation` variant inexplicably (from my POV) changes some things, but doesn't get rid of `python~libxml2` further down the tree.
- On the other hand, adding `^python libxml2==True` to the `+latex ` variant leads to a
```
==> Error: Cyclic dependency detected between 'gettext' and 'python'
    Consider changing variants to avoid the cycle
```

### Information on your system

* **Spack:** 0.20.0.dev0
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-zen2
* **Concretizer:** clingo

### Additional information

Since this is a concretizer issue and the build/installation runs without explicit errors, the build logs aren't particularly pertinent; but here are the build logs for the variant I was originally building (`py-matplotlib+fonts+latex+image+movies+animation`) whereupon I noticed the issue:
[spack-build-env.txt](https://github.com/spack/spack/files/10428590/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/10428591/spack-build-out.txt)
Package maintainer @adamjstewart 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by adamjstewart** [comment]
@alalazo this is an interesting concretizer situation, I'm surprised our current concretizer even allows libxml2 to appear in the DAG twice.

**Comment by rkevk** [comment]
I'm still looking into what exactly happened here and it seems that the two competing python builds were actually installed prior to concretizing `py-matplotlib`. I'm currently checking if I can get rid of the problem by removing `python~libxml2` and reinstalling all of its first-order dependents for `python+libxml2`.

**Comment by adamjstewart** [comment]
Ah, interesting. Yes, when I use `--fresh`, I only get one copy of libxml2 in the spec. So this isn't really a concretization bug or a problem with the dependencies of py-matplotlib, just an interesting quirk that causes things to be reused when they could be reinstalled to simplify the DAG.

**Comment by rkevk** [comment]
So, I'm not sure to what extent this is the root cause, but down the line something seems to go wrong with `texlive`—which is a dependency for `py-matplotlib+latex`—and its dependency `cairo`:
In `package.py` for `texlive`, the dependency is listed as follows:
```    depends_on("cairo+X", when="@2019:")```
But if I install `cairo@1.16.0+X+fc+ft+gobject+pdf~png~svg` according to the spec given below, which should satisfy the dependency requirement, then the concretizer still doesn't like that existing spec and wants to install a separate `cairo@1.16.0+X+fc+ft+gobject+pdf~png~svg` which introduces the competing python install as a dependency.

The point I'm trying to make here is: Why is `texlive` insisting on getting a different version of `cairo` than the existing and supposedly sufficient one, and in particular, why does `texlive` want `cairo ^python~libxml2` despite this never appearing in its `package.py`?

Spec files:
[spack-spec-cairo.txt](https://github.com/spack/spack/files/10429180/spack-spec-cairo.txt)
[spack-spec-texlive.txt](https://github.com/spack/spack/files/10429181/spack-spec-texlive.txt)

**Comment by rkevk** [comment]
I have more information:

In the dependency tree of `texlive`, we find the following:
```
 -   texlive@20220321%gcc@9.4.0 build_system=autotools scheme=small arch=linux-ubuntu20.04-zen2
     [...]
 -       ^ghostscript@9.56.1%gcc@9.4.0~tesseract build_system=autotools arch=linux-ubuntu20.04-zen2
 -           ^gtkplus@2.24.32%gcc@9.4.0~cups~strip build_system=meson buildtype=debugoptimized default_library=shared patches=4e8d441 arch=linux-ubuntu20.04-zen2
              [...]
 -               ^librsvg@2.51.0%gcc@9.4.0~doc build_system=autotools arch=linux-ubuntu20.04-zen2
 -                   ^libcroco@0.6.13%gcc@9.4.0 build_system=autotools arch=linux-ubuntu20.04-zen2
 -                       ^gtk-doc@1.33.2%gcc@9.4.0 build_system=autotools patches=2ba1c03 arch=linux-ubuntu20.04-zen2
 -                           ^docbook-xml@4.3%gcc@9.4.0 build_system=generic arch=linux-ubuntu20.04-zen2
 -                           ^docbook-xsl@1.78.1%gcc@9.4.0 build_system=generic arch=linux-ubuntu20.04-zen2
 -                           ^itstool@2.0.7%gcc@9.4.0 build_system=autotools arch=linux-ubuntu20.04-zen2
 -                           ^libxslt@1.1.33%gcc@9.4.0+crypto~python build_system=autotools arch=linux-ubuntu20.04-zen2
```
In the package files of `itstool`, we see:
```
    depends_on("libxml2+python", type=("build", "run"))
```
(and in `libxslt`, `depends_on("libxml2+python", when="+python")` is given, but as far as I can tell, `libxslt+python` is never requested, so that probably doesn't matter).

This establishes where the requirement `libxml2+python` comes from. Now the essential clash with the existing `python+libxml2` install comes from these lines in the package file of `libxml2`:
```
    # avoid cycle dependency for concretizer
    with when("+python"):
        depends_on("python+shared~libxml2")
        # A note about python versions: libxml 2.10.1 (and presumably earlier) has
        # a bug in its configure script that fails to properly parse python
        # version strings with more than one character for the minor version.
        depends_on("python@:3.9", when="@:2.10.1")
```
In other words, requesting `libxml2+python` will reinstall `python` if the existing `python` is `+libxml2`. I understand the idea behind "avoid cycle dependency", ~~but does the `~libxml2` really have to be specified for the cycle to be avoided?~~ *(Edit: No, but if you leave it out, the concretizer still chooses that variant and won't accept the other one because of the theoretical cycle that it creates)* And if there is already an existing `python` install, isn't it possible to bootstrap from that without getting a cycle?

And finally, is `libxml2+python` really required for `itstool` (@agoodLANL)? I know nothing about ITS Tool, but this whole issue could probably be avoided if it worked with `libxml2~python`...

**Comment by rkevk** [comment]
Seems to be at least somewhat related to https://github.com/spack/spack/pull/20116.

Ultimately, it seems that the issue also could have been prevented if the default option for `python` had been `~libxml2` instead, since—as far as I can tell—nothing depends on `python+libxml2`, but `libxml2+python` *does* depend on `python~libxml2`, and having to install both for no reason is not a sensible default. So at the very least, I'd advocate for changing the default variant of `python` (especially since `+libxml2` doesn't seem to be a particular important or generally necessary variant to me).
